### PR TITLE
task/11504 Make changes to workflows_engine

### PR DESCRIPTION
**Pull request by kenvac**
Adds the feature of populate attribute for components

### CODE REVIEW

File: src/core/containers.py [hunk 1]
```diff
@@ -34,3 +35,28 @@ def __init__(self, flow_name, task_name):
...
+
+class Populate(Container):
+    __slots__ = ["path", "validators", "value"]
+
+    def __init__(self, validators, path=None, value=None, **kwargs):
+        super().__init__(**kwargs)
+        self.path = path
+        self.validators = validators
+        self.value = value
+        if value and path:
```
Comment (nocturnalastro) [inline src/core/containers.py]: could you move this stuff into  ```_validate_args``` or something like that and call it before setting the values?

File: src/core/schema/components/populate.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    "$id": "http://udes.io./jsonschema/workflows/components/populate.json",
+    "title": "display_data",
+    "description": "Workflows populate attribute for component",
+    "type": "object",
+    "definitions": {
+        "validators": {
+            "type": "array",
+            "items": [{ "type": "string" }]
+        },
+        "path": { "type": "string" },
```
Comment (nocturnalastro) [inline src/core/schema/components/populate.json]: this is probably overkill as there not particularly complex nor are they likely to become so

File: src/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class InvalidArguments(Exception):
+    """ Unwanted arguments.
+    Example: When you try to use 'path' and 'value' on input field where only one is allowed."""
+
+    def __init__(self, msg):
```
Comment (nocturnalastro) [inline src/exceptions.py]: just use pass here no need to redefine ```__init__```

File: src/core/containers.py (hunk 1 above)
Comment (kenvac) [reply src/core/containers.py]: done

File: src/core/schema/components/populate.json (hunk 2 above)
Comment (kenvac) [reply src/core/schema/components/populate.json]: simplified :)

File: src/exceptions.py (hunk 3 above)
Comment (kenvac) [reply src/exceptions.py]: done

File: src/core/containers.py [hunk 4]
```diff
@@ -23,14 +20,39 @@ def as_dict(self):
...
 
     def as_dict(self):
         return {"flow": self.flow, "task": self.task}
+
+
+class Populate(Container):
+    __slots__ = ["path", "validators", "value"]
+
+    def _validate_args(self, path, value):
+        if value and path:
```
Comment (nocturnalastro) [inline src/core/containers.py]: value could be 0 so ```is not None``` is what you want to check

**Comment by nocturnalastro** [comment]
Unit tests, please :) e.g  https://github.com/unipartdigital/workflows_engine/blob/master/src/core/tests/tasks/test_condition.py

File: src/core/containers.py (hunk 4 above)
Comment (kenvac) [reply src/core/containers.py:37]: done

Comment (kenvac) [reply src/core/containers.py]: done

File: src/core/containers.py [hunk 5]
```diff
@@ -23,14 +20,39 @@ def as_dict(self):
...
 
     def as_dict(self):
         return {"flow": self.flow, "task": self.task}
+
+
+class Populate(Container):
+    __slots__ = ["path", "validators", "value"]
+
+    def _validate_args(self, path, value):
+        if (value and path) is not None:
```
Comment (nocturnalastro) [inline src/core/containers.py]: this should probably be ```value is not None and path is not None``` because ```(0 and None) is not None ``` is `True`

**Review by nocturnalastro — CHANGES_REQUESTED** [review]

File: src/core/schema/components/populate.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    "title": "populate",
+    "description": "Workflows populate attribute for component",
+    "type": "object",
+    "properties": {
+      "validators": { "type": "array", "items": [{ "type": "string" }] },
+      "path": { "type": "string" },
+      "value": { "type": "integer" }
+    },
+    "required": ["validators"],
+    "oneOf": [{ "required": ["path"] }, { "required": ["value"] }]
```
Comment (nocturnalastro) [inline src/core/schema/components/populate.json]: can you add `  "additionalProperties": false` as well please :)

**Review by nocturnalastro — CHANGES_REQUESTED** [review]

File: src/core/containers.py (hunk 5 above)
Comment (kenvac) [reply src/core/containers.py]: done

File: src/core/schema/components/populate.json (hunk 6 above)
Comment (kenvac) [reply src/core/schema/components/populate.json]: done

File: src/core/tests/components/test_input_populate.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    """ Test populate exceptions raised by wrong parameters """
+    # Populate class with all parameters
+    with pytest.raises(InvalidArguments):
+        Populate(
+            validators=[greater_than_zero()],
+            path="$.pick_details.quants[?(@.product.barcode == 'product')].qty",
+            value=12,
+        )
+    # Populate class without path and value parameters
+    with pytest.raises(InvalidArguments):
```
Comment (nocturnalastro) [inline src/core/tests/components/test_input_populate.py:37]: Split into two tests otherwise, it would take at least two runs to know they're both broken.

Comment (kenvac) [reply src/core/tests/components/test_input_populate.py:37]: done

File: src/exceptions.py (hunk 3 above)
Comment (kdwyer) [inline src/exceptions.py]: Nit: the empty `__init__` is unnecessary - the class statement + docstring is enough to make a valid class declaration.

**Review by kdwyer — APPROVED** [review]

File: src/exceptions.py (hunk 3 above)
Comment (kenvac) [reply src/exceptions.py]: done :)

**Review by nocturnalastro — APPROVED** [review]

**Pull request merged**